docs/design-proposals/high-availability.md

```diff
@@ -0,0 +1,209 @@
+| `cc-shim`    | yes    | container | runtime (vc)         | no          | n/a                  | n/a             | yes ([shim reconnect PR]) | retry                         | no                  | n/a                      |
+| hypervisor   | yes    | pod       | runtime (vc)         | "yes" [2]   | container manager    | "yes"           | n/a                    | n/a                           | no                  | n/a                      |
+| workload     | maybe  | container | agent                | no          | n/a                  | no              | n/a                    | n/a                           | n/a                 | n/a                      |
+
+
+Key:
+
+- n/a - "not applicable" (does not apply).
+
+- If a value is double-quoted, it connotect incomplete support (meaning "sometimes" or "partially").
+
+- [1] - The runtime is invoked multiple times for some container manager operations, complying with the OCI runtime specification.
+
+- [2] - Can be re-run by the container manager as necessary, within the bounds of the OCI specification.
+
+- [3] - Once https://github.com/clearcontainers/proxy/pull/153 lands.
+
+- [4] - Currently, if the agent exits, the virtual machine is shut down so no restart is possible.
+
+- [5] - Visible in [global log](https://github.com/clearcontainers/runtime#global-logfile) and container manager logs.
+
+Most of the components are indirectly launched by a container manager
+(such as Docker*) which calls the runtime component; if the runtime
+detects any errors, these will be reported back to the container manager
+which will take appropriate action.
+
+### Current Proxy
+
+The proxy is currently being reworked for HA on the [proxy state PR].
+
+### Current Shim
+
+Thanks to the [shim reconnect PR], the shim now reconnects to the proxy should
+the latter die. However, it does not cache signals, input or the last command
+run when disconnected from the proxy.
+
+
+## Plan
+
+### Summary of required component changes
+
+#### Runtime
+
+Make runtime attempt to re-connect to the proxy if it is unable to
+connect initially. This requires changes to the `virtcontainers`
+`ccProxy.connectProxy()` function to make it:
+
+- Reconnect and timeout after a period (like the [shim reconnect PR]).
+
+- Record details of the last command `virtcontainers` attempted to send to the
+  proxy (`hyperstartProxyCmd`) in memory.
+
+- If a proxy reconnect is required, re-send the command.
+
+#### Proxy
+
+No changes.
+
+#### Shim
+
+Make the shim cache all signals, input data and the last command run and
+re-send to the proxy when it reconnects after being disconnected from
+the proxy.
+
+#### Agent
+
+- Handle gracefully the re-connection of the proxy to continue properly
+where we left.
+
+- Buffer all outputs supposed to go through STDOUT/STDERR, so that the
+agent can send them after the proxy re-connection. Specific to IO
+channel.
+
+- Save the last command that got executed while the proxy was crashing,
```

Comment:
Lines 124-132 suggested rewrite: Save the last executed command and its result at the time of the proxy crash. Doing so avoids running the same command a second time after the reconnect. Normally, when the proxy reconnects, the last command prior to the disconnect resends because the result was not received. For this change, the agent needs to compare the command sent by the proxy after the reconnect, against the command saved. If the commands are the same, only the saved result is sent and received by the proxy to avoid running duplicate commands. This change will be triggered by the shim or the runtime during the reconnect.